list_symbols.vala

```diff
@@ -1,21 +1,38 @@
 using LanguageServer;
 
+/**
+ * Used to list all symbols defined in a document, usually for outlining.
+ */
 class Vls.ListSymbols : Vala.CodeVisitor {
     private Vala.SourceFile file;
     private Gee.Deque<DocumentSymbol> containers;
     private Gee.List<DocumentSymbol> top_level_syms;
     private Gee.TreeMap<Range, DocumentSymbol> syms_flat;
     private Gee.List<DocumentSymbol> all_syms;
+    private Gee.HashMap<string, DocumentSymbol> ns_name_to_dsym;
+    Vala.TypeSymbol? str_sym; 
 
     public ListSymbols (Vala.SourceFile file) {
         this.file = file;
         this.top_level_syms = new Gee.LinkedList<DocumentSymbol> ();
         this.containers = new Gee.LinkedList<DocumentSymbol> ();
         this.syms_flat = new Gee.TreeMap<Range, DocumentSymbol> ((r1, r2) => r1.start.compare (r2.start));
         this.all_syms = new Gee.LinkedList<DocumentSymbol> ();
+        this.ns_name_to_dsym = new Gee.HashMap<string, DocumentSymbol> ();
+
+        str_sym = file.context.root.scope.lookup ("string") as Vala.TypeSymbol;
+
+        Vala.CodeContext.push (file.context);
         this.visit_source_file (file);
     }
 
+    /**
+     * Keep code context around until we're done using ListSymbols
+     */
+    ~ListSymbols () {
+        Vala.CodeContext.pop ();
```

Comment:
is this the right place for a pop? i could see this possibly breaking if we had multiple ListSymbols in a function with different contexts, as we don't explicitly call the destructor ourselves. was there an issue with putting it after `visit_source_file`?

Reply:
The reason why I added this is because some CodeNodes call `Vala.CodeContext.get ()` in some of their functions, so we need to keep the current `CodeContext` on the top of the stack until ListSymbols goes out of scope and we're therefore no longer accessing any resulting CodeNodes from ListSymbols. I did the same for FindSymbol.

Reply:
yeah i know its required, but it was this before:

```vala
Vala.CodeContext.push (ctx);
this.visit_source_file (sf);
Vala.CodeContext.pop ();
```

are there CodeContext.get() accesses outside that?

Reply:
> are there CodeContext.get() accesses outside that?

Yes. For example, `Vala.ArrayType.get_resize_method ()`

Reply:
we never call that

Reply:
the only thing we do with a `FindSymbol` instance is call `.result`, so i think we can call pop() right at the end of the constructor

Reply:
`get_size_method()` is called by `get_member ()` which may be called when completing, and could access CodeContext if the size_method is null at the time. There are other examples (like `get_end_method ()` in `SignalType`) of this, but I'm having trouble reproducing the scenarios leading up to this. I think they're only likely in very special cases, such as when there are many undefined symbols.

I think there's no harm to leaving this in here.

Reply:
where do we call get_member()?

Reply:
Well as of the recent commit now we call it in `add_completions_for_signal ()`